### PR TITLE
[BEAM-2333] Register a PTransformTranslator for Combine

### PR DESCRIPTION
**Pull request by tgroh**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
Include the Combine Payload in the Runner API Graph.

Add getCombineFn(AppliedPTransform) to extract the CombineFn from an
arbitrary transform.

Update Pipeline Translation tests to include accumulator coders.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @kennknowles

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/SdkComponentsTest.java [hunk 1]
```diff
@@ -124,6 +124,18 @@ public void leaveCompositeTransform(Node node) {
                   equalTo(windowingStrategies.size()));
             } else {
               transforms.add(node);
+              if (PTransformTranslation.COMBINE_TRANSFORM_URN.equals(
+                  PTransformTranslation.urnForTransformOrNull(node.getTransform()))) {
+                // Combine translation introduces a coder that is not assigned to any PCollection
+                // in the default expansion, and must be explicitly added here.
+                try {
+                  addCoders(
+                      CombineTranslation.getAccumulatorCoder(
```
Comment (kennknowles) [inline runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/SdkComponentsTest.java]: Shouldn't `CombinePayloadTranslator.translate` have the ability to get coders where they need to be?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CombineFnTester.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+  public static <InputT, OutputT> void testCombine(
+      CombineFn<InputT, ?, OutputT> combineFn, OutputT expected, Iterable<InputT> inputs) {
+    checkCombineFn(combineFn, Lists.newArrayList(inputs), expected);
+  }
+
+  // //////////////////////////////////////////////////////////////////////////
+  private static <InputT, AccumT, OutputT> void checkCombineFn(
+      CombineFn<InputT, AccumT, OutputT> fn, List<InputT> input, final OutputT expected) {
+    checkCombineFn(fn, input, Matchers.is(expected));
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CombineFnTester.java]: Should this be `equals`? I think `is` makes it object equality, no?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CombineFnTester.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+  }
+
+  private static <InputT, AccumT, OutputT> void checkCombineFn(
+      CombineFn<InputT, AccumT, OutputT> fn, List<InputT> input, Matcher<? super OutputT> matcher) {
+    checkCombineFnInternal(fn, input, matcher);
+    Collections.shuffle(input);
+    checkCombineFnInternal(fn, input, matcher);
+  }
+
+  private static <InputT, AccumT, OutputT> void checkCombineFnInternal(
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CombineFnTester.java]: Why is this `...Internal` when both methods are private? Maybe there's some other way to disambiguate the methods?

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/SdkComponentsTest.java (hunk 1 above)
Comment (tgroh) [inline runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/SdkComponentsTest.java]: It does; it doesn't show in the default expansion of Combine (because we don't group accumulators) but we do translate it (because it gets tossed in the payload)

I'll update the comment.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CombineFnTester.java (hunk 2 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CombineFnTester.java]: `is` and `equalTo` are aliased; you're thinking of `theInstance`

(This is actually not meant to be part of this PR, and I'm going to revert these changes.)

**Comment by tgroh** [comment]
retest this please

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/SdkComponentsTest.java (hunk 1 above)
Comment (kennknowles) [inline runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/SdkComponentsTest.java]: I see what you mean. At this point the test seems a bit underspecific _and_ a bit overspecified. An issue for another day.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CombineFnTester.java (hunk 2 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CombineFnTester.java]: In the future, I'd prefer `equalTo` because it shares more characters with "equals" while the natural interpretation of "is" to me is object identity. A coder "is" another coder only if they really are the same object.
